### PR TITLE
ddl: fix index length calculation

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve? 
In MySQL 8.0: 
```sql
create table t12 (a varchar(1000) primary key);
Error 1071: Specified key was too long; max key length is 3072 bytes
```

In TiDB master:
```sql
create table t12 (a varchar(1000) primary key);
Query OK, 0 rows affected
```

The default charset is `utf8mb4` with at most 4 bytes per character. However, TiDB assumes 1 byte per character by default, which is incorrect. 

### What is changed and how it works?
- Rewrite `getIndexColumnLength`, taking charsets into consideration.
- Add more tests.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test

Code changes

N/A

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release note

 - ddl: fix the index length calculation logic.

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -161,35 +162,45 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
...
-		// Special case for the bit type.
-		if col.FieldType.Tp == mysql.TypeBit {
-			return (col.Flen + 7) >> 3, nil
+	length := takeIfSpecifiedLength(colLen, col.Flen)
+
+	switch col.Tp {
+	case mysql.TypeBit:
+		return (length + 7) >> 3, nil
+	case mysql.TypeVarchar, mysql.TypeString:
+		for _, d := range charset.GetSupportedCharsets() {
```
Comment (bb7133) [inline ddl/index.go]: I think it's better to use `charset.GetCharsetDesc()` here, this function has been used for column length validation somewhere.

File: ddl/index.go [hunk 2]
```diff
@@ -161,35 +162,45 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
...
+	case mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeBlob, mysql.TypeLongBlob:
+		return length, nil
+	case mysql.TypeTiny, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeDouble, mysql.TypeShort, mysql.TypeFloat:
+		defaultLen := mysql.DefaultLengthOfMysqlTypes[col.Tp]
+		if col.Tp == mysql.TypeFloat && length > 24 {
+			// FLOAT(p) has 4 bytes if 0 <= p <= 24, 8 bytes if 25 <= p <= 53
+			defaultLen = defaultLen * 2
+		}
+		return defaultLen, nil
+
```
Comment (bb7133) [inline ddl/index.go]: Remove this blank line?

File: ddl/index.go (hunk 1 above)
Comment (tangenta) [reply ddl/index.go]: Done. PTAL

File: ddl/index.go (hunk 2 above)
Comment (tangenta) [reply ddl/index.go]: Done. PTAL

File: ddl/db_integration_test.go [hunk 3]
```diff
@@ -1464,19 +1464,11 @@ func (s *testIntegrationSuite3) TestAlterColumn(c *C) {
 	c.Assert(err, NotNil)
 
 	s.tk.MustExec("drop table if exists t")
-	// TODO: fix me, below sql should execute successfully. Currently, the result of calculate key length is wrong.
-	//s.tk.MustExec("create table t1 (a varchar(10),b varchar(100),c tinyint,d varchar(3071),index(a),index(a,b),index (c,d));")
-	s.tk.MustExec("create table t1 (a varchar(10),b varchar(100),c tinyint,d varchar(3068),index(a),index(a,b),index (c,d));")
-	_, err = s.tk.Exec("alter table t1 modify column a varchar(3000);")
-	c.Assert(err, NotNil)
-	c.Assert(err.Error(), Equals, "[ddl:1071]Specified key was too long; max key length is 3072 bytes")
+	s.tk.MustExec("create table t1 (a varchar(10),b varchar(100),c tinyint,d varchar(3071),index(a),index(a,b),index (c,d)) charset = ascii;")
```
Comment (wjhuang2016) [inline ddl/db_integration_test.go]: Why set charset = ascii here?

Comment (tangenta) [reply ddl/db_integration_test.go]: In `ascii`, each character takes only one byte. I think this is what the author means.

File: ddl/index.go [hunk 4]
```diff
@@ -161,35 +162,39 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
...
+		desc, err := charset.GetCharsetDesc(col.Charset)
+		if err != nil {
+			return 0, errUnsupportedCharset.GenWithStackByArgs(col.Charset, col.Collate)
+		}
+		return desc.Maxlen * length, nil
+	case mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeBlob, mysql.TypeLongBlob:
+		return length, nil
+	case mysql.TypeTiny, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeDouble, mysql.TypeShort, mysql.TypeFloat:
+		defaultLen := mysql.DefaultLengthOfMysqlTypes[col.Tp]
+		if col.Tp == mysql.TypeFloat && length > 24 {
```
Comment (wjhuang2016) [inline ddl/index.go]: Please use some meaningful names instead of magic number

Comment (wjhuang2016) [inline ddl/index.go]: Split TypeFloat with other types?

Comment (tangenta) [reply ddl/index.go]: The comment below explains. BTW, it's hard to find a clear and meaningful name in this case. Any suggestion?

Comment (wjhuang2016) [reply ddl/index.go]: Some kinds of "BytesMaxLength"? And what if p > 53?

Comment (wjhuang2016) [reply ddl/index.go]: FloatMaxLength or f32MaxLength, singlePrecisionMaxLength?             
length > 24 will be a double

Comment (tangenta) [reply ddl/index.go]: @wjhuang2016 Done. How about this?

Comment (wjhuang2016) [reply ddl/index.go]: Great

File: ddl/index.go [hunk 5]
```diff
@@ -161,35 +162,39 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
 }
 
 func getIndexColumnLength(col *model.ColumnInfo, colLen int) (int, error) {
-	// Take care of the sum of length of all index columns.
+	length := types.UnspecifiedLength
```
Comment (wjhuang2016) [inline ddl/index.go]: Please add some comments for this function, and point out the effect of the charset.

File: ddl/index.go [hunk 6]
```diff
@@ -161,35 +162,39 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
...
+	case mysql.TypeTiny, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeDouble, mysql.TypeShort:
+		return mysql.DefaultLengthOfMysqlTypes[col.Tp], nil
+	case mysql.TypeFloat:
+		if length <= mysql.MaxFloatPrecisionLength {
+			return mysql.DefaultLengthOfMysqlTypes[mysql.TypeFloat], nil
+		}
+		return mysql.DefaultLengthOfMysqlTypes[mysql.TypeDouble], nil
+	case mysql.TypeDecimal, mysql.TypeNewDecimal:
+		// Decimal using a binary format that packs nine decimal (base 10) digits into four bytes.
+		return (length / 9 * 4) + ((length%9)+1)/2, nil
```
Comment (wjhuang2016) [inline ddl/index.go]: Could you put it into a function? Thus we can understand getIndexColumnLength easily

**Pull request merged**